server/src/main/java/org/candlepin/async/impl/ArtemisJobMessageDispatcher.java

```diff
@@ -0,0 +1,154 @@
+import org.apache.activemq.artemis.api.core.client.ServerLocator;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.inject.Inject;
+
+
+
+/**
+ * The ArtemisJobMessageDispatcher is a job message dispatcher implementation that uses Artemis
+ * to distribute job messages using a single message queue.
+ */
+public class ArtemisJobMessageDispatcher implements JobMessageDispatcher {
+    private static Logger log = LoggerFactory.getLogger(ArtemisJobMessageDispatcher.class);
+
+    private Configuration config;
+    private ObjectMapper objMapper;
+
+    private ClientSessionFactory sessionFactory;
+    private ClientSession session;
+    private ClientProducer producer;
+
+
+    @Inject
+    public ArtemisJobMessageDispatcher(Configuration config, ObjectMapper objMapper) {
+        this.config = config;
+        this.objMapper = objMapper;
+    }
+
+
+    /**
+     * Fetches the current session factory, initializing a new instance as necessary. This should
+     * almost certainly not be used by any method other than getClientSession.
+     *
+     * @throws ActiveMQException
+     *  if an exception occurs while spinning up the ActiveMQ client session factory
+     *
+     * @return
+     *  the ClientSessionFactory instance for this job manager
+     */
+    private ClientSessionFactory getClientSessionFactory() throws Exception {
+        if (this.sessionFactory == null || this.sessionFactory.isClosed()) {
+            log.debug("Creating new ActiveMQ client session factory...");
+
+            ServerLocator locator = ActiveMQClient.createServerLocator(
+                this.config.getProperty(ConfigProperties.ACTIVEMQ_BROKER_URL));
+
+            // TODO: Maybe make this a bit more defensive and skip setting the property if it's
+            // not present in the configuration rather than crashing out?
+            locator.setMinLargeMessageSize(this.config.getInt(ConfigProperties.ACTIVEMQ_LARGE_MSG_SIZE));
+
+            try {
+                this.sessionFactory = locator.createSessionFactory();
+                log.debug("Created new ActiveMQ client session factory: {}", this.sessionFactory);
+            }
+            catch (Exception e) {
+                if (e instanceof ActiveMQException) {
+                    throw (ActiveMQException) e;
+                }
+                else {
+                    throw new RuntimeException(e);
+                }
+            }
+        }
+
+        return this.sessionFactory;
+    }
+
+    /**
+     * Fetches the current client session, creating and intializing a new instance as necessary.
+     *
+     * @return
+     *  the current ClientSession instance for this job manager
+     */
+    ClientSession getClientSession() throws Exception {
+        if (this.session == null || this.session.isClosed()) {
+            log.debug("Creating new ActiveMQ session for async job messages...");
+
+            ClientSessionFactory csf = this.getClientSessionFactory();
```

Comment:
This is where i think the connection would come into play:
`this.connection.getSenderSession()`

Reply:
You could also bundle the session/producer into an object and retrieve those straight from the connection as well. This could be very similar to what we do in EventMessageSender the internal class that we use in EventSinkImpl.

Anyway, something to consider.